delve/torchcallback.py

```diff
@@ -297,47 +300,62 @@ def record_layer_saturation(layer: torch.nn.Module, input, output):
                     training_state = 'train' if layer.training else 'eval'
 
-                    self.logs[f'{training_state}-saturation'][
-                        layer.name] = saturation
-                if 'spectral' not in stats:
-                    if 'eigendist' in stats:
-                        eig_vals = hooks.add_eigen_dist(
-                            layer, eig_vals, layer.forward_iter)
-                    if 'neigendist' in stats:
-                        eig_vals = hooks.add_neigen_dist(
-                            layer, eig_vals, layer.forward_iter)
-                    if 'spectrum' in stats:
-                        eig_vals = hooks.add_spectrum(
-                            layer,
-                            eig_vals,
-                            top_eigvals=5,
-                            n_iter=layer.forward_iter)
-                if 'param_eigenvals' in stats:
-                    layer_param_weight = layer.state_dict()['weight']
-                    u, s, v = layer_param_weight.svd()
-                    param_eigenvals = hooks.add_param_eigenvals(
-                        layer, s, top_eigvals=5, n_iter=layer.forward_iter)
+                    self.logs[f'{training_state}-saturation'][layer.name] = saturation
+                elif 'cov' in stats:
+                    training_state = 'train' if layer.training else 'eval'
+
+                    if len(activations_batch.shape) == 4:  # conv layer (B x C x H x W)
+                        if self.conv_method == 'median':
+                            activations_batch = np.median(activations_batch, axis=(2, 3))  # channel median
+                        elif self.conv_method == 'max':
+                            activations_batch = np.max(activations_batch, axis=(2, 3))  # channel median
+                        elif self.conv_method == 'mean':
+                            activations_batch = np.mean(activations_batch, axis=(2, 3))
+
+                    if layer.name in self.logs[f'{training_state}-saturation']:
+                        self.logs[f'{training_state}-saturation'][layer.name].update(activations_batch)
+                    else:
+                        self.logs[f'{training_state}-saturation'][layer.name] = CovarianceMatrix()
+                        self.logs[f'{training_state}-saturation'][layer.name]._init_internals(activations_batch)
+
 
         layer.register_forward_hook(record_layer_saturation)
+
+
+    def add_saturations(self):
+        """
+        Computes saturation and saves all stats
+        :return:
+        """
+        for key in self.logs.keys():
+            train_sats = []
+            val_sats = []
+            if '-saturation' in key:
+                for layer_name in self.logs[key].keys():
+                    if not self.ignore_layer_names is None and self.ignore_layer_names == layer_name:
+                        continue
+                    sat = compute_saturation(self.logs[key][layer_name]._cov_mtx, thresh=.99)
```

Comment:
the threshold should be straightforward to configure from outside the class

Reply:
done